docs/recipes/immutability-helpers.md

```diff
@@ -0,0 +1,190 @@
+# Immutability Helpers
+
+Redux is a tiny pattern that represents states as immutable objects. Redux was originally designed for React. Most Redux concepts, such as pure functions, are centered around the React ecosystem. Nowadays Redux is not directly related to React.
+
+The cornerstone of Redux is immutability. Immutabilty is an amazing pattern to minimise unpredictable behaviour in our code. We're not going to cover functional programming in this article. However we're going to look at very useful packages that are called "immutability helpers". Most developers have to deal with, so called, "deep objects" and most important follow the immutability concept, when it comes to changing the value of some deeply nested property.
+
+## immer
+
+`immer` is the most popular library that allows you to work with immutable objects in a more convenient way. Given the following code:
+
+```ts
+export interface Task {
+  title: string;
+  dates: {
+    startDate: string;
+    dueDate: string;
+  };
+}
+
+export interface TrelloStateModel {
+  tasks: {
+    [taskId: string]: Task;
+  };
+}
+
+@State<TrelloStateModel>({
+  name: 'trello',
+  defaults: {
+    tasks: {}
+  }
+})
+export class TrelloState {}
+```
+
+Let's imagine that we're faced with the task of changing the `dueDate` property:
+
+```ts
+export class UpdateDueDate {
+  static readonly type = '[Trello] Update due date';
+  constructor(public taskId: string, public dueDate: string) {}
+}
+```
+
+Let's see how we would implement the `updateDueDate` action handler:
+
+```ts
+export class TrelloState {
+
+  @Action(UpdateDueDate)
+  updateDueDate(ctx: StateContext<TrelloStateModel>, action: UpdateDueDate) {
+    ctx.setState(state => ({
+      tasks: {
+        ...state.tasks,
+        [action.taskId]: {
+          ...state.tasks[action.taskId],
+          dates: {
+            ...state.tasks[action.taskId].dates,
+            dueDate: action.dueDate
+          }
+        }
+      }
+    }));
+  }
+
+}
+```
+
+Such code is complicated to maintain and accompany. It's not self-descriptive and will be a problem for upcoming developers. Let's see how we could re-write this code with the help of `immer`:
+
+```ts
+import { produce } from 'immer';
+
+export class TrelloState {
+
+  @Action(UpdateDueDate)
+  updateDueDate(ctx: StateContext<TrelloStateModel>, action: UpdateDueDate) {
+    const state = produce(ctx.getState(), draft => {
+      draft.tasks[action.taskId].dates.dueDate = action.dueDate;
+    });
+
+    ctx.setState(state);
+  }
+
+}
+```
+
+Oh, have you noticed it's less code and looks much better. From the `immer` repisitory:
+> Using Immer is like having a personal assistant; he takes a letter (the current state) and gives you a copy (draft) to jot changes onto. Once you are done, the assistant will take your draft and produce the real immutable, final letter for you (the next state).
+
```

Comment:
In addition to the example above I think that we should show that you can also use immer's `produce` function as a state operator like this:
```
import { produce } from 'immer';
export class TrelloState {
  @Action(UpdateDueDate)
  updateDueDate(ctx: StateContext<TrelloStateModel>, action: UpdateDueDate) {
    ctx.setState(produce(draft => {
      draft.tasks[action.taskId].dates.dueDate = action.dueDate;
    }));
  }
}
```

Reply:
OK, I will add 2 ways of using `immer`